tests/schema/tests.py

```diff
@@ -2676,6 +2676,39 @@ class Meta:
         with self.assertRaises(IntegrityError):
             Tag.objects.create(title="bar", slug="foo")
 
+    def test_unique_with_fk_and_condition_multiple_times(self):
+        """
+        Tests adding and removing a unique constraint to ForeignKey
+        with a condition multiple times.
+        """
+        with connection.schema_editor() as editor:
+            editor.create_model(Author)
+            editor.create_model(Book)
+        constraint = UniqueConstraint(
+            "author",
+            condition=Q(title__in=["tHGttG", "tRatEotU"]),
+            name="book_author_condition_uniq",
+        )
+        table = Book._meta.db_table
+
+        for i in range(2):
+            # Add constraint.
+            with connection.schema_editor() as editor:
+                editor.add_constraint(Book, constraint)
+                sql = constraint.create_sql(Book, editor)
+            constraints = self.get_constraints(table)
+            if connection.features.supports_partial_indexes:
+                self.assertIn(constraint.name, constraints)
+                self.assertIs(constraints[constraint.name]["unique"], True)
+                self.assertIn("WHERE %s IN" % editor.quote_name("title"), str(sql))
+            else:
+                self.assertNotIn(constraint.name, constraints)
+                self.assertIsNone(sql)
+            # Remove constraint.
+            with connection.schema_editor() as editor:
+                editor.remove_constraint(Book, constraint)
+            self.assertNotIn(constraint.name, self.get_constraints(table))
```

Comment:
I'd assert the number of constraints instead of checking that removing constraint doesn't crash, e.g.:
```python
    def test_remove_ignored_unique_constraint_not_create_fk_index(self):
        with connection.schema_editor() as editor:
            editor.create_model(Author)
            editor.create_model(Book)
        constraint = UniqueConstraint(
            "author",
            condition=Q(title__in=["tHGttG", "tRatEotU"]),
            name="book_author_condition_uniq",
        )
        table = Book._meta.db_table
        # Add unique constraint.
        with connection.schema_editor() as editor:
            editor.add_constraint(Book, constraint)
            sql = constraint.create_sql(Book, editor)
        old_constraints = self.get_constraints_for_column(
            Book,
            Book._meta.get_field("author").column,
        )   
        # Remove unique constraint.
        with connection.schema_editor() as editor:
            editor.remove_constraint(Book, constraint)
        new_constraints = self.get_constraints_for_column(
            Book,
            Book._meta.get_field("author").column,
        )   
        # Redundant foreign key index is not added.
        self.assertEqual(
            len(old_constraints) - 1
            if connection.features.supports_partial_indexes
            else len(old_constraints),
            len(new_constraints),
        )   
```

Reply:
That makes much more sense, thanks!